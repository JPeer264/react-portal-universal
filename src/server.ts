import * as ReactDOMServer from "react-dom/server";
import { load }  from "cheerio";
import { flushUniversalPortals } from "./index";

export function appendUniversalPortals(html: string) {
  const $ = load(html);
  flushUniversalPortals().forEach(([children, selector]) => {
    const markup = ReactDOMServer.renderToStaticMarkup(children);
    $(markup).attr("data-react-universal-portal", "").appendTo((selector as any))
  });
  return $.html();
}
