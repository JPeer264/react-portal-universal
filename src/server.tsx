import * as React from "react";
import * as ReactDOMServer from "react-dom/server";
import { load }  from "cheerio";
import { PortalManager, ChildrenSelectorTuple } from "./PortalManager";

export class ServerPortal {
  portals: ChildrenSelectorTuple[];

  constructor() {
    this.portals = [];
    this.appendUniversalPortals = this.appendUniversalPortals.bind(this);
  }

  collectPortals(children: React.ReactNode) {
    return <PortalManager portals={this.portals}>{children}</PortalManager>;
  }

  appendUniversalPortals(html: string) {
    if (!this.portals.length) {
      return html;
    }

    const $ = load(html);
    this.portals.forEach(([children, selector]) => {
      const markup = ReactDOMServer.renderToStaticMarkup(children);
      $(markup).attr("data-react-universal-portal", "").appendTo((selector as any))
    });

    // it's important to flush one and only one time per render
    this.portals.length = 0;

    return $.html({ decodeEntities: false });
  }
}
