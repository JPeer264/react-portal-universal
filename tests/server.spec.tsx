import * as React from "react";
import { createUniversalPortal, flushUniversalPortals } from "../src";
import { appendUniversalPortals } from "../src/server";

describe("server-side", () => {
  beforeEach(() => {
    flushUniversalPortals();
  });

  describe("appendUniversalPortals", () => {
    it("adds static mackup from portals to rendered HTML", () => {
      const html = `
        <html>
          <head>
            <meta charset="utf-8" />
          </head>
          <body>
            <div id="root"></div>
          </body>
        </html>
      `;
      createUniversalPortal(<title>Hello, World!</title>, "head");
      createUniversalPortal(<h1>Hello, World!</h1>, "body");
      const withStaticMarkup = appendUniversalPortals(html);

      expect(withStaticMarkup.includes("<title>Hello, World!</title>"));
      expect(withStaticMarkup.includes("<h1>Hello, World!</h1>"));
      expect(withStaticMarkup).toMatchSnapshot();
    });
  });
});
