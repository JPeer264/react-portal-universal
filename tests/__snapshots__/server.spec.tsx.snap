// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server-side appendUniversalPortals adds static mackup from portals to rendered HTML 1`] = `
"<html><head>
            <meta charset=\\"utf-8\\">
          <title>Hello, World!</title></head>
          <body>
            <div id=\\"root\\"></div>
          
        
      <h1>Hello, World!</h1></body></html>"
`;
